nixos/modules/services/logging/promtail.nix

```
@@ -10,15 +10,29 @@ let
 
   allowPositionsFile = !lib.hasPrefix "/var/cache/promtail" positionsFile;
   positionsFile = cfg.configuration.positions.filename;
+
+  configFile = if cfg.configFile != null
+                   then cfg.configFile
+                   else prettyJSON cfg.configuration;
+
 in {
   options.services.promtail = with types; {
     enable = mkEnableOption "the Promtail ingresser";
 
-
     configuration = mkOption {
       type = (pkgs.formats.json {}).type;
       description = ''
         Specify the configuration for Promtail in Nix.
+        This option is incompatible with `configFile`.
```

In my opinion, it would be nice if the description made it clear that if `configFile` is present, `configuration` is ignored.
And I would write out the names of the options.

```suggestion
        This option will be ignored if `services.promtail.configFile` is defined.
```